torchvision/prototype/transforms/_auto_augment.py

```diff
@@ -0,0 +1,296 @@
+            extra_kwargs=dict(angle=0.0, translate=[0, 0], scale=1.0),
+            takes_interpolation_kwargs=True,
+        ),
+        "ShearY": AutoAugmentDispatcher(
+            F.affine,
+            magnitude_fn=lambda magnitude: dict(shear=[0, math.degrees(magnitude)]),
+            extra_kwargs=dict(angle=0.0, translate=[0, 0], scale=1.0),
+            takes_interpolation_kwargs=True,
+        ),
+        "TranslateX": AutoAugmentDispatcher(
+            F.affine,
+            magnitude_fn=lambda magnitude: dict(translate=[int(magnitude), 0]),
+            extra_kwargs=dict(angle=0.0, scale=1.0, shear=[0.0, 0.0]),
+            takes_interpolation_kwargs=True,
+        ),
+        "TranslateY": AutoAugmentDispatcher(
+            F.affine,
+            magnitude_fn=lambda magnitude: dict(translate=[0, int(magnitude)]),
+            extra_kwargs=dict(angle=0.0, scale=1.0, shear=[0.0, 0.0]),
+            takes_interpolation_kwargs=True,
+        ),
+        "Rotate": AutoAugmentDispatcher(F.rotate, magnitude_fn=lambda magnitude: dict(angle=magnitude)),
+        "Brightness": AutoAugmentDispatcher(
+            F.adjust_brightness, magnitude_fn=lambda magnitude: dict(brightness_factor=1.0 + magnitude)
+        ),
+        "Color": AutoAugmentDispatcher(
+            F.adjust_saturation, magnitude_fn=lambda magnitude: dict(saturation_factor=1.0 + magnitude)
+        ),
+        "Contrast": AutoAugmentDispatcher(
+            F.adjust_contrast, magnitude_fn=lambda magnitude: dict(contrast_factor=1.0 + magnitude)
+        ),
+        "Sharpness": AutoAugmentDispatcher(
+            F.adjust_sharpness, magnitude_fn=lambda magnitude: dict(sharpness_factor=1.0 + magnitude)
+        ),
+        "Posterize": AutoAugmentDispatcher(F.posterize, magnitude_fn=lambda magnitude: dict(bits=int(magnitude))),
+        "Solarize": AutoAugmentDispatcher(F.solarize, magnitude_fn=lambda magnitude: dict(threshold=magnitude)),
+        "AutoContrast": AutoAugmentDispatcher(F.autocontrast),
+        "Equalize": AutoAugmentDispatcher(F.equalize),
+        "Invert": AutoAugmentDispatcher(F.invert),
+    }
+
+    def get_transforms_meta(self, image_size: Tuple[int, int]) -> Iterator[Tuple[str, float]]:
+        raise NotImplementedError
+
+    def get_params(self, sample: Any) -> Dict[str, Any]:
+        image = Query(sample).image_for_size_and_channels_extraction()
+
+        fill = self.fill
+        if isinstance(fill, (int, float)):
+            fill = [float(fill)] * image.num_channels
+        elif fill is not None:
+            fill = [float(f) for f in fill]
+
+        return dict(interpolation=self.interpolation, fill=fill)
+
+    def forward(self, *inputs: Any, params: Optional[Dict[str, Any]] = None) -> Any:
+        sample = inputs if len(inputs) > 1 else inputs[0]
+        params = params or self.get_params(sample)
+
+        for transform_id, magnitude in self.get_transforms_meta(Query(sample).image_size()):
```

Comment:
AutoAugmentation methods should be structured exactly like any other transform. In that sense, can we refactor this have a `_transform` where we do all the operators and remove `get_transforms_meta`?

Reply:
We cannot. `Transform._transform` already gets passed the non-container elements of the inputs. For the auto augment transforms, we need to apply multiple transforms to the whole input. To illustrate, this PR proposes

```py
for transform in transforms:
    for non_container_element in inputs:
        transform(non_container_element)
```

while your proposal would mean

```py
for non_container_element in inputs:
    for transform in transforms:
        transform(non_container_element)
```

We reserved the option to override the `forward` of a transform for specifically cases like this, when the common structure does not fit the use case.

Reply:
OK, we should try to simplify the `forward` even if this means implementing it separately for each transform. To me, the `get_transforms_meta` is not clear what is doing and I'm afraid it's not generic enough to cover all cases (see AugMix for instance #5411 which is a substantial departure from the existing methods). The less abstraction the better.

Reply:
> OK, we should try to simplify the `forward` even if this means implementing it separately for each transform.

I'm against that. Most transforms need the default behavior. In fact, the auto augment base class is the only transform that deviates from this scheme. I wouldn't make it more complicated for all other cases just to not have a single (or maybe very few) exceptions in the future. 

> To me, the `get_transforms_meta` is not clear what is doing

I have a hard time to figure out what you want me to document on these PRs and what not. Maybe this is a naming thing (as I said, I'm not really deep into the literature here so maybe my names are wrong), but this simply yields a tuple with the name of the op to apply and the magnitude. It is only used in `forward` to achieve the loop I detailed in my comment above. 

> The less abstraction the better.

I somewhat disagree. I we can simplify our implementation significantly by having a thin abstraction (in this case a single method), I don't think we should strive for no abstraction at all. I'm aware that researchers do crazy things, but our code is also not set in stone. If we later find that some parts are not general enough, we can revise them. It is impossible to foresee the future, so I don't want to maintain a lot of boilerplate code for a maybe upcoming scenario.

> I'm afraid it's not generic enough to cover all cases (see AugMix for instance #5411 which is a substantial departure from the existing methods)

That being said, it seems the abstraction I built is already no general enough for a very tangible scenario and thus I'm ok to remove it. My suggestion is to do this when #5411 is merged.

Reply:
@pmeier This is not a theoretical debate; I've provided you a clear example where the proposed abstraction won't work. I'm just waiting for the authors of the paper to confirm the default hyper-params other than this the AugMix PR is ready to merge. Thus there is no point adding this abstraction now just to remove it later. So here we either need to make the abstraction clear and generic enough to be useful or eliminate it.